webapp/src/Controller/Team/SubmissionController.php

```diff
@@ -60,47 +61,136 @@ public function createAction(Request $request, ?Problem $problem = null): Respon
 
-        $form->handleRequest($request);
+        $formPaste = $this->formFactory
+            ->createBuilder(SubmitProblemPasteType::class, $data)
+            ->setAction($this->generateUrl('team_submit'))
+            ->getForm();
 
-        if ($form->isSubmitted() && $form->isValid()) {
+        $formUpload->handleRequest($request);
+        $formPaste->handleRequest($request);
+        if ($formUpload->isSubmitted() || $formPaste->isSubmitted()) {
             if ($contest === null) {
                 $this->addFlash('danger', 'No active contest');
             } elseif (!$this->dj->checkrole('jury') && !$contest->getFreezeData()->started()) {
                 $this->addFlash('danger', 'Contest has not yet started');
             } else {
-                /** @var Problem $problem */
-                $problem = $form->get('problem')->getData();
-                /** @var Language $language */
-                $language = $form->get('language')->getData();
-                /** @var UploadedFile[] $files */
-                $files      = $form->get('code')->getData();
-                if (!is_array($files)) {
-                    $files = [$files];
+                $problem = null;
+                $language = null;
+                $files = [];
+                $entryPoint = null;
+                $message = '';
+        
+                if ($formUpload->isSubmitted() && $formUpload->isValid()) {
+                    $problem = $formUpload->get('problem')->getData();
+                    $language = $formUpload->get('language')->getData();
+                    $files = $formUpload->get('code')->getData();
+                    if (!is_array($files)) {
+                        $files = [$files];
+                    }
+                    $entryPoint = $formUpload->get('entry_point')->getData() ?: null;
+                } elseif ($formPaste->isSubmitted() && $formPaste->isValid()) {
+                    $problem = $formPaste->get('problem')->getData();
+                    $language = $formPaste->get('language')->getData();
+                    $codeContent = $formPaste->get('code_content')->getData();
+        
+                    if ($codeContent == null || empty(trim($codeContent))) {
+                        $this->addFlash('danger', 'No code content provided.');
+                        return $this->redirectToRoute('team_index');
+                    }
+        
+                    $tempDir = sys_get_temp_dir();
+                    $tempFileName = sprintf(
+                        'submission_%s_%s_%s.%s',
+                        $user->getUsername(),
+                        $problem->getName(),
+                        date('Y-m-d_H-i-s'),
+                        $language->getExtensions()[0]
+                    );
+                    $tempFileName = preg_replace('/[^a-zA-Z0-9_.-]/', '_', $tempFileName);
+
+                    if ($language->getExtensions()[0] == 'java' || $language->getExtensions()[0] == 'kt') {
```

Comment:
Instead of hardcoding Java and Kotlin here, use the `$language->getRequireEntryPoint()` instead.

Similarly, use `$language->getEntryPointDescription()` for a description to display.

Reply:
We probably want to use what we built here so that the entry point works the same regardless of the upload methos: https://github.com/DOMjudge/domjudge/blob/main/webapp/src/Form/Type/SubmitProblemType.php#L66

Reply:
Sorry, I don't understand how to use $language->getRequireEntryPoint(). For example, in the case of Java, when using paste_code, the entry_point might still be no. In this case, how can I use this to determine the file name standard?

Reply:
I think the remark here is that instead of listing the current languages which need an entrypoint, use that function @meisterT linked and use it from there?

Reply:
> I think the remark here is that instead of listing the current languages which need an entrypoint, use that function @meisterT linked and use it from there?

The function should be checking whether the language in the form requires an entrypoint, which seems to be different from my current situation, right?

Reply:
Why would it be different? You check for 2 languages here and determine the entrypoint for those. If there are other languages like this we would also want to pick them up. People can add more languages besides those 2.

Reply:
@vmcj 
<img width="850" alt="截圖 2024-11-16 晚上8 10 24" src="https://github.com/user-attachments/assets/e7492d59-3d57-4d3a-afba-35a83bd2794d">
I'm not sure if my understanding is correct, but are the methods **$language->getRequireEntryPoint()** and **$language->getEntryPointDescription()** linked to the part I highlighted in the image?

If so, there might be an issue.

Reply:
For example, in the case of Java, it can be configured to not require an entry point. In this scenario, under the original upload mode, the uploaded file itself becomes the entry point. However, due to the characteristics of the Java language, the file name must match the main class name.

In the case of the paste code mode, regardless of whether an entry point is required or not, the user must be allowed to specify the main class name. Otherwise, I won’t be able to save the file correctly, which will result in a compile error.